### PR TITLE
Starknet foundry unit testing

### PR DESCRIPTION
**Pull request by aagbotemi**
This is the implementation of Starknet foundry unit testing using snforge

### CODE REVIEW

**Comment by tekkac** [comment]
run `scarb fmt` on the code :)

**Comment by aagbotemi** [comment]
I ran `scarb fmt` and `scarb build` on the code, and it was successful.

File: tests/erc7498_unit_test.cairo [hunk 1]
```diff
@@ -0,0 +1,475 @@
+use core::clone::Clone;
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: We should avoid cloning

File: tests/erc7498_unit_test.cairo [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+use core::traits::TryInto;
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: I think this one is importent automatically in the prelude

File: tests/erc7498_unit_test.cairo [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+    let second_redeem_contract_address = redeem_contract.deploy(@calldata).unwrap();
+    let second_redeem_token = IERC721RedeemablesMixinDispatcher {
+        contract_address: second_redeem_contract_address
+    };
+    let second_redeem_token_safe = IERC721RedeemablesMixinSafeDispatcher {
+        contract_address: second_redeem_contract_address
+    };
+
+    let receive_contract = declare("ERC721RedemptionMintable");
+    calldata.append_serde(test_address());
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: Could you explain the use of `test_address()` in this context?

File: tests/erc7498_unit_test.cairo [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+
+fn CAMPAIGN_URI() -> ByteArray {
+    "https://example.com/campaign"
+}
+
+fn CAMPAIGN_URI_NEW() -> ByteArray {
+    "https://example.com/new/campaign"
+}
+
+fn setup() -> (
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: Could this be imported from the other existing test file?

File: tests/erc7498_unit_test.cairo [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+    };
+
+    let consideration = array![
+        ConsiderationItem {
+            item_type: ItemType::ERC721_WITH_CRITERIA,
+            token: redeem_contract_address,
+            identifier_or_criteria: 0,
+            start_amount: 1,
+            end_amount: 1,
+            recipient: BURN_ADDRESS()
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo:37]: Why BURN_ADDRESS here?

File: tests/erc7498_unit_test.cairo [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+        }
+    ];
+
+    (offer, consideration)
+}
+
+#[test]
+fn supports_interface() {
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: This test is already in the other test file I believe.

File: tests/erc7498_unit_test.cairo [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+    ) =
+        setup();
+    assert!(redeem_token.supports_interface(IERC7498_ID));
+}
+
+#[test]
+fn test_mint_token_success() {
+    let (redeem_contract_address, redeem_token, _, _, _, _, _, _, _) = setup();
+
+    redeem_token.set_approval_for_all(redeem_contract_address, true);
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: Who is calling set_approval_for_all?
Might need a prank from a user here.

File: tests/erc7498_unit_test.cairo [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+    redeem_token.mint(OWNER(), TOKEN_ID);
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: Some checks on the minted token might be missing here.

File: tests/erc7498_unit_test.cairo [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+    let (redeem_contract_address, redeem_token, _, _, _, _, _, _receive_token, _) = setup();
+
+    start_prank(CheatTarget::One(redeem_contract_address), ACCOUNT1());
+
+    redeem_token.set_approval_for_all(redeem_contract_address, true);
+    redeem_token.mint(OWNER(), TOKEN_ID);
+}
+
+#[test]
+#[should_panic()]
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: An expected panic message is missing here

File: tests/erc7498_unit_test.cairo [hunk 10]
```diff
@@ -0,0 +1,475 @@
...
+#[test]
+fn test_burn_token_success() {
+    let (_, redeem_token, _, _, _, _, _, _receive_token, _) = setup();
+
+    redeem_token.mint(test_address(), TOKEN_ID);
+    redeem_token.burn(TOKEN_ID);
+}
+
+#[test]
+#[should_panic()]
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: should catch the exact error message here

File: tests/erc7498_unit_test.cairo [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+
+    redeem_token.set_approval_for_all(redeem_contract_address, true);
+    redeem_token.mint(test_address(), TOKEN_ID);
+
+    let (offer, consideration) = offer_and_consideration(
+        receive_contract_address, redeem_contract_address
+    );
+
+    let mut offer_array = array![];
+    offer_array.append(offer);
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo:54]:
```suggestion
    let mut offer_array = array![offer];
```

File: tests/erc7498_unit_test.cairo [hunk 12]
```diff
@@ -0,0 +1,475 @@
...
+
+    let requirements = array![
+        CampaignRequirements { offer: offer_array.span(), consideration: consideration.span() }
+    ];
+
+    let timestamp: u32 = get_block_timestamp().try_into().unwrap();
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: I think timestamps should be u64, might need a change in the lib.

File: tests/erc7498_unit_test.cairo [hunk 13]
```diff
@@ -0,0 +1,475 @@
...
+        _,
+        _,
+        _,
+        receive_contract_address,
+        receive_token,
+        _
+    ) =
+        setup();
+
+    redeem_token.set_approval_for_all(redeem_contract_address, true);
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: You don't need approval here

File: tests/erc7498_unit_test.cairo [hunk 14]
```diff
@@ -0,0 +1,475 @@
...
+    redeem_token.mint(test_address(), TOKEN_ID);
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: mint_redemption already mints a NFT...

File: tests/erc7498_unit_test.cairo [hunk 15]
```diff
@@ -0,0 +1,475 @@
...
+
+    let (offer, consideration) = offer_and_consideration(
+        receive_contract_address, redeem_contract_address
+    );
+
+    receive_token.mint_redemption(CAMPAIGN_ID, test_address(), offer, consideration.span());
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo:85]: checks post mint missing

File: tests/erc7498_unit_test.cairo [hunk 16]
```diff
@@ -0,0 +1,475 @@
...
+}
+
+#[test]
+#[should_panic()]
```
Comment (tekkac) [inline tests/erc7498_unit_test.cairo]: error missing

**Review by tekkac — CHANGES_REQUESTED** [review]
Thanks for your patience on the review.
Some additional rework is needed.
- mint test on the token aren't needed, what we'd like to unit test is the EIP7498 methods.
- some approvals aren't needed.
- most post method call checks are missing. <-- We want to check that the logic of the redemption is working here.
 - boundary checks are missing. <-- What happens when no token are left to redeem, etc.

Please try to not reuse code and import test methods you need if they exist.

Hope this is helpful.

File: tests/erc7498_unit_test.cairo (hunk 5 above)
Comment (aagbotemi) [reply tests/erc7498_unit_test.cairo:37]: The recipient can actually be any address, meanwhile i have updated the recipient address.

File: tests/erc7498_unit_test.cairo (hunk 3 above)
Comment (aagbotemi) [reply tests/erc7498_unit_test.cairo]: it has been updated

File: tests/erc7498_unit_test.cairo (hunk 1 above)
Comment (aagbotemi) [reply tests/erc7498_unit_test.cairo]: Alright

**Comment by aagbotemi** [comment]
Hi @tekkac, I've done the unit testing and resolved the issues you raised.

**Comment by tekkac** [comment]
I'm closing this one as it has been superseded by the current Seaport issues.
Checks after executing the methods are still missing and it's not mergable so I'll close the PR.
If you're still interested in contributing, you could check with the contributors on the current Seaport issues if they want to split and pick up some of their work.

**Pull request closed without merging**